sklearn/cluster/k_means_.py

```diff
@@ -685,7 +685,9 @@ def _init_centroids(X, k, init, random_state=None, x_squared_norms=None,
         seeds = random_state.permutation(n_samples)[:k]
         centers = X[seeds]
     elif hasattr(init, '__array__'):
-        centers = init
+        # ensure that the centers have the same dtype as X
+        # this is a requirement of fused types of cython
+        centers = np.array(init, dtype=X.dtype)
```

Comment:
Yeah but I still change [this line](https://github.com/yenchenlin/scikit-learn/blob/9e81c041de1f4c5e410c99221c490d9c0b6851e6/sklearn/cluster/k_means_.py#L308) to `init = check_array(init, dtype=X.dtype.type, copy=True)` since it is more consistent to keep it as the same datatype as `X`, and also it makes no harm to precision! 😃